API/Tensorflow/native_libs/src/ops/Const.h

```diff
@@ -16,7 +16,7 @@ template <TF_DataType DataTypeLabel>
 class Const : public Operation<DataTypeLabel> {
 public:
     explicit Const(std::shared_ptr<Tensor<DataTypeLabel>> tensor) : value(std::move(tensor)) {
-    	hash = (value->hash());
+    	hash = (value->hash() ^ const_salter.next_salt());
```

Comment:
This requires a comment documenting why we decide to do that.

Reply:
Although mathematically two consts sharing the same value are the same objects, we decided to distinguish them. Look at the following piece of code, from the regression example:
```
[a,b] = train . fold [0.0, 0.0] gradientDescent
```
where the step function creates constTensor form those zeros, calculates gradient of error function and calls batchEval. We fold over a list of values, using stochastic gradient descent, and our accumulator is initially a zero scalar. But this way, the step function would treat both of the inputs, wrapped into Const Op, **as the same object**, although we did not intend to treat them that way. That would result in returning **always the same values** for **a** and **b**. If we changed the code into something like
```
[a,b] = train . fold [0.0, 0.01] gradientDescent,
```
it would lead to totally different (and much better) results (e.g, returned a and b would be different, and proper)! Generally, it is not a wise way to differentiate by Const; the more mathematical way is to use Placeholder, or unimplemented yet Variable instead. But because it **is** possible in the API, we have to distinguish consts having the same values but being different objects in some way, and xoring its hash with some uniformly-distributed salt is a way to achieve this.